Design.md

```diff
@@ -32,71 +32,73 @@ Here are the IELE design decisions described from the perspective of differences
 
 * Because words are now explicitly signed, the SDIV, SMOD, SLT, and SGT instructions are removed, and DIV, MOD, LT, and GT are replaced with signed versions.
-* An EXPMOD instruction is added to perform exponentiation modulo a particular base, and the MODEXP precompiled contract is removed.
-* SIGNEXTEND is changed to convert an N-byte twos-complement representation of a signed number into its signed value.
-* a TWOS instruction is added to convert a signed integer into an N-byte twos-complement representation of the number.
-* Because integers are arbitrary-precision, the integer operand of BYTE now counts from the least-significant byte instead of the most-significant.
+* An `expmod` instruction is added to perform exponentiation modulo a particular base, and the MODEXP precompiled contract is removed.
+* `sext` (corresponding to EVM's SIGNEXTEND) is changed to convert an N-byte twos-complement representation of a signed number into its signed value.
+* a `twos` instruction is added to convert a signed integer into an N-byte twos-complement representation of the number.
+* Because integers are arbitrary-precision, the integer operand of `byte` (corresponding to EVM's BYTE) now counts from the least-significant byte instead of the most-significant.
 
-## Contract Creation
+## Instructions and Precompiled Contracts
+
+Unlike EVM, where each instruction is just an opcode operating in the values found in the top of the stack, IELE instructions follow a syntax similar to LLVM: each instruction is an opcode followed by some arguments in the form of registers. If the instruction produces a result, it is formed as an assignment where the LHS contains the register(s) to hold the result(s) and the RHS contains the opcode and the arguments. Other changes with respect to EVM include:
 
-Unlike EVM, for ease of verification we disallow creation of dynamic code for dynamically created contracts, and instead restrict dynamically-created-contracts to code that consists of some previously validated contract.
+* IELE has one integer comparison instruction, namely `cmp`, that accepts various predicates. In addition to the predicates `lt`, `gt`, and `eq`, which have corresponding EVM instructions, the predicates `le`, `ge`, and `ne` are also supported.
+* Various EVM instructions that query local and/or network state (e.g. GAS, BALANCE, etc.) have been replaced by IELE builtins that can be called using the `call` instruction (e.g. `%balance = call @iele.balance(%bank.account)`). The names of all IELE builtins start with the prefix `@iele.`, following a convention similar to the LLVM intrinsics. All names for globals (registers and/or functions) starting with this prefix are reserved: A contract is malformed if it contains user-defined registers and/or functions with names starting with `@iele.`.
+* Precompiled contracts in IELE are also invoked using corresponding builtins (e.g. `@iele.ecpairing`).
```

Comment:
Are the precompiled contracts accessed as a local call, or do they need to be written as a remote call that passes an appropriate amount of gas?

Reply:
This is a good question. Currently they can only be called as an account call, and I think that is the correct and intended way to call them. However such a call would look kind of bad in textual IELE because it would require to load the constant 1 as the address of the precompiled account:

```
%addr = 1
%val = ...
%gas = ...
%res  = call @iele.sha256 at %addr ( %len, %data ) send %val, gaslimit %gas
```
Maybe @dwightguth can comment on that too.